### PR TITLE
feat: Publish updates to datasets when publishing a collection

### PR DESCRIPTION
**Pull request by Bento007**
### Reviewers
**Functional:** @MDunitz 

**Readability:** @seve 

---
Work for #1145

rdev deployment https://tsmith-revision-frontend.rdev.single-cell.czi.technology 

## Changes
- moved revision related publishing tests to test_revision
- add logic to copy revision datasets to published dataset
- add logic to delete tombstones dataset when publishing a revision
- published tombstones are not returned in collection details.
- remove Dataset.delete_explorer_cxg_object_from_s3 and replacing with Dataset.asset_deletion()
- add check on backend to verify at least one dataset exists in a private collection before publishing.
- add commit flag to functions that commit to database. This gives us better control over when changes are committed to the database. This will make it easier to roll back changes if publish fails.
- add after session commit hook to delete s3_assets
- moving s3_client and s3_resource to s3_bucket
- deleting a refreshed dataset, restores the original dataset to the revision.

## Definition of Done (from ticket)
- A revision can publish a refreshed dataset
- The old dataset info will no longer be accessible
- The new dataset is one revision higher than the old dataset after publishing
- The new dataset has published=True after publishing
- The new dataset has original_uud=None after publishing

## Known Issues
- additional work https://github.com/chanzuckerberg/corpora-data-portal/issues/1249
  -  grey out publish button if no datasets are in a collection
  - add pop up before publishing a revision
  - add toast after publishing a revision.

### CODE REVIEW

File: backend/corpora/common/entities/collection.py [hunk 1]
```diff
@@ -182,9 +182,35 @@ def publish(self):
...
-            if dataset.original_id:
-                "skip modified datasets"
-                continue  # TODO: expand to support tombstone and refresh corpora-data-portal/1177
+            ds = Dataset(dataset)
+            ds_original_id = ds.original_id
+            if ds_original_id:
+                original = Dataset.get(self.session, ds_original_id)
+                ds_tombstone = ds.tombstone
+                ds_revision = ds.revision
+                if (public_collection.id, public_collection.visibility) == (
```
Comment (MDunitz) [inline backend/corpora/common/entities/collection.py:191]: There is a lot going on in this if statement, what if you pulled it out into a separate check function?

Comment (Bento007) [reply backend/corpora/common/entities/collection.py:191]: that would be a lot of variables going into that functions. How about this 
```
if all([public_collection.id == original.collection_id,
          public_collection.visibility == original.collection_visibility,
          ds_tombstone or ds_revision > original.revision]):
```

File: tests/unit/backend/chalice/api_server/test_revisions.py [hunk 2]
```diff
@@ -290,6 +294,35 @@ def verify_datasets(self, actual_body, expected_dataset_ids):
...
+        self.app.delete(f"/dp/v1/datasets/{rev_dataset_id}", self.headers)
+        self.publish_collection(self.rev_collection.id)
+        dataset = Dataset.get(self.session, pub_dataset.id, include_tombstones=True)
+        self.assertTrue(dataset.tombstone)
+        for s3_object in published_s3_objects:
+            self.assertS3FileDoesNotExist(*s3_object)
+
+    def test__with_revision_with_refreshed_datasets__OK(self):
+        """"publish a revision with refreshed datasets"""
+        """publish a revision with delete datasets"""
```
Comment (MDunitz) [inline tests/unit/backend/chalice/api_server/test_revisions.py:321]:
```suggestion
```

**Review by Bento007 — COMMENTED** [review]
- add commit flag to line 172 of collection.py to prevent early commit.
- don't delete the datset_assets until after the DB changes have been commited

**Review by seve — APPROVED** [review]
Readability ✅

**Review by MDunitz — APPROVED** [review]
lgtm - one thought it might be worth adding a test of the rollback functionality to make sure if something goes wrong at the end of publishing a revision it is able to rollback to the previous state

**Pull request merged**